File: examples/integration/billing_outbox.mjs
#!/usr/bin/env node
// Synthetic caller outbox for one ordinary billing accept request.
// Usage: node billing_outbox.mjs LEDGER BILLING_DIR CUSTOMER SOURCE EVENT_JSON OUTBOX_DIR
// Run one caller process per outbox on a private, durable local filesystem.
import { spawnSync } from 'node:child_process';
import { closeSync, existsSync, fsyncSync, lstatSync, mkdirSync, openSync, readFileSync, realpathSync, renameSync, writeSync } from 'node:fs';
import { dirname, join } from 'node:path';
import { isDeepStrictEqual } from 'node:util';

function fail(message) { throw new Error(message); }
function syncDir(directory) {
  const fd = openSync(directory, 'r');
  try { fsyncSync(fd); } finally { closeSync(fd); }
}
function syncFile(path) {
  const fd = openSync(path, 'r+');
  try { fsyncSync(fd); } finally { closeSync(fd); }
}
function saveNew(path, bytes, directory) {
  const fd = openSync(path, 'wx', 0o600);
  try {
    let written = 0;
    while (written < bytes.length) {
      const count = writeSync(fd, bytes, written, bytes.length - written);
      if (count <= 0) fail('outbox write made no progress');
      written += count;
    }
    fsyncSync(fd);
  } finally { closeSync(fd); }
  syncDir(directory);
}
function same(a, b) { return isDeepStrictEqual(a, b); }
function runJson(args) {
  const result = spawnSync(args[0], args.slice(1), { encoding: 'utf8', maxBuffer: 128 * 1024 * 1024 });
  if (result.error) fail(`ledger failed to start: ${result.error.message}`);
  let value;
  try { value = JSON.parse(result.stdout); } catch { fail(`ledger returned non-JSON output (exit ${result.status})`); }
  return [result.status, value];
}

if (process.argv.length !== 8) fail('usage: node billing_outbox.mjs LEDGER BILLING_DIR CUSTOMER SOURCE EVENT_JSON OUTBOX_DIR');
const [ledger, suppliedInstallation, customer, source, eventFile, outbox] = process.argv.slice(2);
const installation = realpathSync(suppliedInstallation);
const eventBytes = readFileSync(eventFile);
let event;
try { event = JSON.parse(eventBytes.toString('utf8')); } catch { fail('event must be strict JSON'); }
if (event?.schema !== 'ledger-event/1' || event.customer !== customer || typeof event.id !== 'string' || !event.id ||
    typeof event.operation_id !== 'string' || !event.operation_id) fail('event must contain stable id and operation_id');

if (!existsSync(outbox)) {
  mkdirSync(outbox, { mode: 0o700 });
}
const outboxStat = lstatSync(outbox);
if (!outboxStat.isDirectory() || outboxStat.isSymbolicLink())
  fail('outbox must be a private directory, not a symlink');
if (outboxStat.mode & 0o077) fail('outbox must have no group or other permission bits');
syncDir(dirname(outbox));
const marker = join(outbox, 'installation.path');
const installed = Buffer.from(installation + '\n' + customer + '\n' + source + '\n');
if (existsSync(marker)) {
  if (!readFileSync(marker).equals(installed)) fail('outbox belongs to another canonical installation or customer/source scope');
} else {
  saveNew(marker, installed, outbox);
}
syncFile(marker);
syncDir(outbox);
const request = join(outbox, 'request.json');
if (existsSync(request)) {
  const saved = readFileSync(request);
  if (!saved.equals(eventBytes)) fail('event bytes differ from the pending request; retain the original IDs and bytes');
  const prior = JSON.parse(saved.toString('utf8'));
  if (prior.id !== event.id || prior.operation_id !== event.operation_id) fail('saved request IDs differ');
} else {
  saveNew(request, eventBytes, outbox);
}
syncFile(request);
syncDir(outbox);

const [code, response] = runJson([ledger, 'billing', '--directory', installation, 'accept', '--customer', customer, '--source', source, request, '--json']);
if (code === 8) {
  console.error('Outcome unknown. Keep request.json pending and retry this identical file.');
  process.exit(8);
}
if (code !== 0 || !['accepted', 'duplicate'].includes(response?.status)) {
  console.error(`Accept not acknowledged (exit ${code}): ${JSON.stringify(response)}`);
  process.exit(code || 1);
}
const receipt = response.receipt;
const target = receipt?.body?.target;
if (receipt?.kind !== 'base-acceptance' || typeof receipt.id !== 'string' || !receipt.id ||
    typeof target !== 'string' || !target) fail('receipt ID or target is missing');
const [explainCode, history] = runJson([ledger, 'billing', '--directory', installation, 'explain', '--customer', customer, target, '--json']);
if (explainCode !== 0 || history?.schema !== 'ledger-billing-statement/2' || history.complete !== true)
  fail('complete target explanation unavailable; request remains unacknowledged');
const matches = history.entries?.filter(entry => entry.target === target &&
  same(entry.receipt, receipt) && entry.receipt.id === receipt.id &&
  entry.records?.filter(record => record.kind === 'event' &&
    record.body?.data?.operation_id === event.operation_id).length === 1) ?? [];
if (matches.length !== 1) fail('receipt does not match the retained original operation');

const receiptPath = join(outbox, 'receipt.json');
if (existsSync(receiptPath)) {
  if (!same(JSON.parse(readFileSync(receiptPath, 'utf8')), receipt))
    fail('saved receipt differs; investigate before acknowledging');
  syncFile(receiptPath);
  syncDir(outbox);
} else {
  const temporary = join(outbox, 'receipt.json.pending');
  const bytes = Buffer.from(JSON.stringify(receipt) + '\n');
  if (existsSync(temporary)) {
    if (!readFileSync(temporary).equals(bytes)) fail('incomplete or different pending receipt; investigate before acknowledging');
  } else {
    saveNew(temporary, bytes, outbox);
  }
  syncFile(temporary);
  syncDir(outbox);
  renameSync(temporary, receiptPath);
  syncFile(receiptPath);
  syncDir(outbox);
}
console.log(`Acknowledged ${response.status} with original receipt ${receipt.id}`);

File: examples/integration/billing_outbox.py
#!/usr/bin/env python3
"""Synthetic caller outbox for one ordinary billing accept request.

Usage: python3 billing_outbox.py LEDGER BILLING_DIR EVENT_JSON OUTBOX_DIR
The outbox and installation require a private, durable local filesystem.
Run one caller process per outbox; do not use real customer data here.
"""

import json
import os
from pathlib import Path
import stat
import subprocess
import sys


def stop(message):
    raise SystemExit(message)


def sync_dir(path):
    fd = os.open(path, os.O_RDONLY)
    try:
        os.fsync(fd)
    finally:
        os.close(fd)


def sync_file(path):
    fd = os.open(path, os.O_RDWR)
    try:
        os.fsync(fd)
    finally:
        os.close(fd)


def save_new(path, data):
    fd = os.open(path, os.O_WRONLY | os.O_CREAT | os.O_EXCL, 0o600)
    try:
        with os.fdopen(fd, "wb") as stream:
            stream.write(data)
            stream.flush()
            os.fsync(stream.fileno())
    finally:
        sync_dir(path.parent)


def save_receipt(path, receipt):
    data = (json.dumps(receipt, sort_keys=True, separators=(",", ":")) + "\n").encode()
    if path.exists():
        if json.loads(path.read_bytes()) != receipt:
            stop("saved receipt differs; investigate before acknowledging")
        sync_file(path)
        sync_dir(path.parent)
        return
    temporary = path.with_name(path.name + ".pending")
    if temporary.exists():
        if temporary.read_bytes() != data:
            stop("incomplete or different pending receipt; investigate before acknowledging")
    else:
        save_new(temporary, data)
    sync_file(temporary)
    sync_dir(path.parent)
    os.replace(temporary, path)
    sync_file(path)
    sync_dir(path.parent)


def run_json(argv):
    result = subprocess.run(argv, capture_output=True, text=True)
    try:
        value = json.loads(result.stdout)
    except json.JSONDecodeError:
        stop(f"ledger returned non-JSON output (exit {result.returncode})")
    return result.returncode, value


def verified_receipt(ledger, installation, operation_id, receipt):
    if not isinstance(receipt, dict) or receipt.get("kind") != "base-acceptance":
        stop("missing base-acceptance receipt")
    target = receipt.get("body", {}).get("target")
    if not isinstance(target, str) or not target or not isinstance(receipt.get("id"), str):
        stop("receipt ID or target is missing")
    code, history = run_json([ledger, "billing", "--directory", installation, "explain", target, "--json"])
    if code != 0 or history.get("schema") != "ledger-billing-statement/1" or history.get("complete") is not True:
        stop("complete target explanation unavailable; request remains unacknowledged")
    matches = [entry for entry in history.get("entries", [])
               if entry.get("target") == target and entry.get("receipt") == receipt
               and entry["receipt"].get("id") == receipt["id"]
               and len([record for record in entry.get("records", [])
                        if record.get("kind") == "event"
                        and record.get("body", {}).get("data", {}).get("operation_id") == operation_id]) == 1]
    if len(matches) != 1:
        stop("receipt does not match the retained original operation")
    return receipt


def main():
    if len(sys.argv) != 5:
        stop(__doc__)
    ledger, installation, event_file, outbox_name = sys.argv[1:]
    installation = str(Path(installation).resolve(strict=True))
    source = Path(event_file).read_bytes()
    try:
        event = json.loads(source)
    except (ValueError, UnicodeDecodeError):
        stop("event must be strict JSON")
    if not isinstance(event, dict) or event.get("schema") != "ledger-event/1" or any(
        not isinstance(event.get(key), str) or not event[key] for key in ("id", "operation_id")
    ):
        stop("event must contain stable id and operation_id")

    outbox = Path(outbox_name)
    if not outbox.exists():
        outbox.mkdir(mode=0o700)
    mode = outbox.stat().st_mode
    if outbox.is_symlink() or not stat.S_ISDIR(mode):
        stop("outbox must be a private directory, not a symlink")
    if mode & 0o077:
        stop("outbox must have no group or other permission bits")
    sync_dir(outbox.parent)
    marker = outbox / "installation.path"
    installed = (installation + "\n").encode("utf-8")
    if marker.exists():
        if marker.read_bytes() != installed:
            stop("outbox belongs to another canonical installation path")
    else:
        save_new(marker, installed)
    sync_file(marker)
    sync_dir(outbox)
    request = outbox / "request.json"
    if request.exists():
        saved = request.read_bytes()
        if saved != source:
            stop("event bytes differ from the pending request; retain the original IDs and bytes")
        saved_event = json.loads(saved)
        if (saved_event["id"], saved_event["operation_id"]) != (event["id"], event["operation_id"]):
            stop("saved request IDs differ")
    else:
        save_new(request, source)
    sync_file(request)
    sync_dir(outbox)

    code, response = run_json([ledger, "billing", "--directory", installation, "accept", str(request), "--json"])
    if code == 8:
        print("Outcome unknown. Keep request.json pending and retry this identical file.", file=sys.stderr)
        return 8
    if code != 0 or response.get("status") not in ("accepted", "duplicate"):
        print(f"Accept not acknowledged (exit {code}): {response}", file=sys.stderr)
        return code or 1
    receipt = verified_receipt(ledger, installation, event["operation_id"], response.get("receipt"))
    save_receipt(outbox / "receipt.json", receipt)
    print(f"Acknowledged {response['status']} with original receipt {receipt['id']}")
    return 0


if __name__ == "__main__":
    sys.exit(main())
